include/seastar/util/program-options.hh

```diff
@@ -463,6 +466,27 @@ public:
     void set_value(T value) { do_set_value(std::move(value), false); }
 };
 
+template <typename T = std::monostate>
+class updateable_value : public value<T> {
+    noncopyable_function<void(const updateable_value&)> _update;
+    virtual void do_mutate_live(options_mutator& mutator) override {
+        value<T>::do_mutate(mutator);
+        _update(*this);
```

Comment:
This will unconditionally call `_update()` for all values, regardless of whether their value changed or not. Not sure if this is a problem.
We could make `do_mutate()` return `bool` and return `true` if value changed. We don't need to try very hard to detect changes, it is enough if we detect those cases where a value wasn't updated at all (it is still default). Most options are left as default anyway and are never updated.

Reply:
Good point. I like the latter approach so the plain

```c++
    virtual void do_mutate_live(options_mutator& mutator) override {
        value<T>::do_mutate(mutator);
        if (!_defaulted) {
            _update(*this);
        }
       ...
```

should work well enough

Reply:
Note that we have to detect the case where an option that previously had a value, is now defaulted (option was removed from settings file altogether). This won't detect that.
Actually, I think the current code will not even update such an option at all as `mutator.visit_value(val)` will return false for such values. When doing a live update and the visitor returns false from `visit_value` (meaning it has no value for this setting), we should reset the option back to default.

So we would invoke the updater in two cases: 
* `visit_value()` returned `true`
* `visit_value()` returned `false` and `_defaulted` was `false` (the value is being reset to default).